// vue/src/core/instance/render-helpers/render-static.js
/* @flow */

/**
 * Runtime helper for rendering static trees.
 */
// 辅助渲染函数：用于渲染静态树(双向绑定需要相应式更新，这部分变化时需要重新生成，这部分是动态的,另一部分就是静态的不会变的，那么就缓存起来复用)
export function renderStatic (
  index: number,
  isInFor: boolean
): VNode | Array<VNode> {
  const cached = this._staticTrees || (this._staticTrees = [])
  let tree = cached[index]
  // if has already-rendered static tree and not inside v-for,
  // we can reuse the same tree.
  // 如果已经渲染过静态树，同时不是在v-for中，那么直接重用
  if (tree && !isInFor) {
    return tree
  }
  // otherwise, render a fresh tree.
  // 否则的话渲染一棵新树
  tree = cached[index] = this.$options.staticRenderFns[index].call(
    this._renderProxy,
    null,
    this // for render fns generated for functional component templates
  )
  // 生成树后把树标记为静态树
  markStatic(tree, `__static__${index}`, false)
  return tree
}

/**
 * Runtime helper for v-once.
 * Effectively it means marking the node as static with a unique key.
 */
// 渲染辅助函数，用于处理v-once指令
export function markOnce (
  tree: VNode | Array<VNode>,
  index: number,
  key: string
) {
  markStatic(tree, `__once__${index}${key ? `_${key}` : ``}`, true)
  return tree
}
// 分情况标记静态节点，是数组的话就为判断每个元素然后标记，不是的话直接标记
function markStatic (
  tree: VNode | Array<VNode>,
  key: string,
  isOnce: boolean
) {
  if (Array.isArray(tree)) {
    for (let i = 0; i < tree.length; i++) {
      if (tree[i] && typeof tree[i] !== 'string') {
        markStaticNode(tree[i], `${key}_${i}`, isOnce)
      }
    }
  } else {
    markStaticNode(tree, key, isOnce)
  }
}
// 把节点标记为静态节点
function markStaticNode (node, key, isOnce) {
  node.isStatic = true
  node.key = key
  node.isOnce = isOnce
}
