// vue/src/platforms/web/util/element.js
/* @flow */

import { inBrowser } from 'core/util/env'
import { makeMap } from 'shared/util'
// 定义常量,命名空间表
export const namespaceMap = {
  svg: 'http://www.w3.org/2000/svg',
  math: 'http://www.w3.org/1998/Math/MathML'
}
// 生成一个判断是否是html标签的元素
export const isHTMLTag = makeMap(
  'html,body,base,head,link,meta,style,title,' +
  'address,article,aside,footer,header,h1,h2,h3,h4,h5,h6,hgroup,nav,section,' +
  'div,dd,dl,dt,figcaption,figure,picture,hr,img,li,main,ol,p,pre,ul,' +
  'a,b,abbr,bdi,bdo,br,cite,code,data,dfn,em,i,kbd,mark,q,rp,rt,rtc,ruby,' +
  's,samp,small,span,strong,sub,sup,time,u,var,wbr,area,audio,map,track,video,' +
  'embed,object,param,source,canvas,script,noscript,del,ins,' +
  'caption,col,colgroup,table,thead,tbody,td,th,tr,' +
  'button,datalist,fieldset,form,input,label,legend,meter,optgroup,option,' +
  'output,progress,select,textarea,' +
  'details,dialog,menu,menuitem,summary,' +
  'content,element,shadow,template,blockquote,iframe,tfoot'
)

// this map is intentionally selective, only covering SVG elements that may
// contain child elements.
// 此映射是有意选择的，只覆盖可能包含子元素的svg元素
// 生成一个判断是否是svg的函数
export const isSVG = makeMap(
  'svg,animate,circle,clippath,cursor,defs,desc,ellipse,filter,font-face,' +
  'foreignObject,g,glyph,image,line,marker,mask,missing-glyph,path,pattern,' +
  'polygon,polyline,rect,switch,symbol,text,textpath,tspan,use,view',
  true
)
// 判断是否是pre标签
export const isPreTag = (tag: ?string): boolean => tag === 'pre'
// 判断是否是保留标签
export const isReservedTag = (tag: string): ?boolean => {
  return isHTMLTag(tag) || isSVG(tag)
}
// 获取标签所属命名空间
export function getTagNamespace (tag: string): ?string {
  if (isSVG(tag)) {
    return 'svg'
  }
  // basic support for MathML
  // note it doesn't support other MathML elements being component roots
  // MathML的基本支持。注意，它不支持将其他MathML作为组件根
  if (tag === 'math') {
    return 'math'
  }
}

const unknownElementCache = Object.create(null)
// 判断是否是未识别标签
export function isUnknownElement (tag: string): boolean {
  // 不是浏览器环境，返回true
  /* istanbul ignore if */
  if (!inBrowser) {
    return true
  }
  // 是保留标签，返回false
  if (isReservedTag(tag)) {
    return false
  }
  tag = tag.toLowerCase()
  // 如果之前已经判断过存在缓存里了，不用再判断直接返回结果
  /* istanbul ignore if */
  if (unknownElementCache[tag] != null) {
    return unknownElementCache[tag]
  }
  // 如果是浏览器环境，需要判断的标签又不是保留标签，而之前又没有被判断过,
  // 那么以此标签创建一个dom节点，继续判断
  const el = document.createElement(tag)
  // 如果标签中含有连字符，判断其构造函数
  // 本身存在的节点都有特定的构造函数，比如div是window.HTMLDivElement,a是window.HTMLAnchorElement等，
  // 不存在的标签会调用window.HTMLUnknownElement或window.HTMLElement，故以此来做判断
  if (tag.indexOf('-') > -1) {
    // http://stackoverflow.com/a/28210364/1070244
    return (unknownElementCache[tag] = (
      el.constructor === window.HTMLUnknownElement ||
      el.constructor === window.HTMLElement
    ))
  // 如果标签中不存在连字符，那么把el字符串化，不存在的标签其值为"[object HTMLUnknownElement]"
  } else {
    return (unknownElementCache[tag] = /HTMLUnknownElement/.test(el.toString()))
  }
}
// 判断是否是input标签的类型
export const isTextInputType = makeMap('text,number,password,search,email,tel,url')
