// _vue@2.5.16@vue/src/platforms/web/runtime/directives/model.js
/**
 * Not type checking this file because flow doesn't like attaching
 * properties to Elements.
 */

import { isTextInputType } from 'web/util/element'
import { looseEqual, looseIndexOf } from 'shared/util'
import { mergeVNodeHook } from 'core/vdom/helpers/index'
import { warn, isIE9, isIE, isEdge } from 'core/util/index'

// selectionchange事件在文档上的当前文本选择被改变时触发
// document.onselectionchange = function() { console.log("Selection changed!"); };
// 在谷歌下，无论是鼠标左键点击不同元素、右键、拖动选中文本、input框值的变动都会触发
// 在火狐下，只有左键点击不同元素、拖动选中文本会触发，右键、input框值的变动不会触发
// ie9以上，左键点击不同元素（点击相同的input也会）、拖动选中文本、input值变动会触发，右键不会触发
// 故selectionchange相当于时当前文档的焦点变更或内容更改触发
/* istanbul ignore if */
if (isIE9) {
  // http://www.matts411.com/post/internet-explorer-9-oninput/
  // 大概就是ie9下的input事件会有毛病，所以如果是ie9的话，用selectchange事件来捕获并触发input事件
  document.addEventListener('selectionchange', () => {
    const el = document.activeElement
    if (el && el.vmodel) {
      trigger(el, 'input')
    }
  })
}
// compositionstart事件触发于一段文字的输入之前
// compositionend事件将被触发于文本段落的组成完成或取消之时
const directive = {
  inserted (el, binding, vnode, oldVnode) {
    if (vnode.tag === 'select') {
      // #6903
      if (oldVnode.elm && !oldVnode.elm._vOptions) {
        mergeVNodeHook(vnode, 'postpatch', () => {
          directive.componentUpdated(el, binding, vnode)
        })
      } else {
        setSelected(el, binding, vnode.context)
      }
      el._vOptions = [].map.call(el.options, getValue)
    } else if (vnode.tag === 'textarea' || isTextInputType(el.type)) {
      el._vModifiers = binding.modifiers
      if (!binding.modifiers.lazy) {
        el.addEventListener('compositionstart', onCompositionStart)
        el.addEventListener('compositionend', onCompositionEnd)
        // Safari < 10.2 & UIWebView doesn't fire compositionend when
        // switching focus before confirming composition choice
        // this also fixes the issue where some browsers e.g. iOS Chrome
        // fires "change" instead of "input" on autocomplete.
        el.addEventListener('change', onCompositionEnd)
        /* istanbul ignore if */
        if (isIE9) {
          el.vmodel = true
        }
      }
    }
  },
  // 
  componentUpdated (el, binding, vnode) {
    if (vnode.tag === 'select') {
      setSelected(el, binding, vnode.context)
      // in case the options rendered by v-for have changed,
      // it's possible that the value is out-of-sync with the rendered options.
      // detect such cases and filter out values that no longer has a matching
      // option in the DOM.
      const prevOptions = el._vOptions
      const curOptions = el._vOptions = [].map.call(el.options, getValue)
      if (curOptions.some((o, i) => !looseEqual(o, prevOptions[i]))) {
        // trigger change event if
        // no matching option found for at least one value
        const needReset = el.multiple
          ? binding.value.some(v => hasNoMatchingOption(v, curOptions))
          : binding.value !== binding.oldValue && hasNoMatchingOption(binding.value, curOptions)
        if (needReset) {
          trigger(el, 'change')
        }
      }
    }
  }
}

function setSelected (el, binding, vm) {
  actuallySetSelected(el, binding, vm)
  /* istanbul ignore if */
  if (isIE || isEdge) {
    setTimeout(() => {
      actuallySetSelected(el, binding, vm)
    }, 0)
  }
}

function actuallySetSelected (el, binding, vm) {
  const value = binding.value
  const isMultiple = el.multiple
  if (isMultiple && !Array.isArray(value)) {
    process.env.NODE_ENV !== 'production' && warn(
      `<select multiple v-model="${binding.expression}"> ` +
      `expects an Array value for its binding, but got ${
        Object.prototype.toString.call(value).slice(8, -1)
      }`,
      vm
    )
    return
  }
  let selected, option
  for (let i = 0, l = el.options.length; i < l; i++) {
    option = el.options[i]
    if (isMultiple) {
      selected = looseIndexOf(value, getValue(option)) > -1
      if (option.selected !== selected) {
        option.selected = selected
      }
    } else {
      if (looseEqual(getValue(option), value)) {
        if (el.selectedIndex !== i) {
          el.selectedIndex = i
        }
        return
      }
    }
  }
  if (!isMultiple) {
    el.selectedIndex = -1
  }
}
// 判断options中的属性是否都与value结构不同
function hasNoMatchingOption (value, options) {
  return options.every(o => !looseEqual(o, value))
}
// 返回option的值value
function getValue (option) {
  return '_value' in option
    ? option._value
    : option.value
}
// 加个标志表明处于文本将要输入的状态
function onCompositionStart (e) {
  e.target.composing = true
}
// 如果是文本输入完成的状态，更改标志表明输入完成，之后触发input事件
function onCompositionEnd (e) {
  // prevent triggering an input event for no reason
  // 防止毫无理由的触发input事件
  if (!e.target.composing) return
  e.target.composing = false
  trigger(e.target, 'input')
}
// 触发事件
function trigger (el, type) {
  // document.createEvent会创建一个指定类型的事件
  const e = document.createEvent('HTMLEvents')
  // initEvent(type, bubbles, cancelable)（已废弃）用来初始化有document.createEvent创建的事件
  // type定义了事件的类型，bubbles决定事件是否向上冒泡，cancelable决定事件的默认动作是否可以被取消
  e.initEvent(type, true, true)
  // 触发事件监听
  el.dispatchEvent(e)
}

export default directive
